### PR TITLE
Uniffi demo qr scan

### PR DESCRIPTION
**Pull request by swaptr**
Merge only after #896 
This PR elaborates the Uniffi demo and API.

### CODE REVIEW

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                            )
+                        )
+                        .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
+                        .build()
+                    imageAnalysis.setAnalyzer(
+                        ContextCompat.getMainExecutor(context),
+                        QrCodeAnalyzer { result ->
+                            code = result
+                            showDialog = true
+                        }
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:115]: IMO it might be cleaner if we do the following:
1. rename `code` to something more specific, e.g. `scannedQRCodeText`
2. make `scannedQRCodeText` a nullable string
3. remove `showDialog`
4. instead of `if (showDialog) { ... }` you can do `scannedQRCodeText?.let { text -> ... }`
5. have the dismiss button have an action that sets `scannedQRCodeText = null`

just lets us remove a state variable, which is usually nice iMO

**Comment by swaptr** [comment]
On generating a new indy profile, the application terminates.
Logcat:

```bash
2023-08-29 13:18:04.298 14352-14418 AndroidRuntime          org.hyperledger.ariesvcx             E  FATAL EXCEPTION: DefaultDispatcher-worker-1
                                                                                                    Process: org.hyperledger.ariesvcx, PID: 14352
                                                                                                    org.hyperledger.ariesvcx.VcxUniFfiException$SerializationException: A serialization error occurred. Check your inputs. More Info: EOF while parsing a value at line 1 column 0
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.read(vcx.kt:915)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.read(vcx.kt:910)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverter$DefaultImpls.liftFromRustBuffer(vcx.kt:151)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterRustBuffer$DefaultImpls.liftFromRustBuffer(vcx.kt:163)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.liftFromRustBuffer(vcx.kt:910)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.liftFromRustBuffer(vcx.kt:910)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterRustBuffer$DefaultImpls.lift(vcx.kt:164)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.lift(vcx.kt:910)
                                                                                                    	at org.hyperledger.ariesvcx.FfiConverterTypeVcxUniFfiError.lift(vcx.kt:910)
                                                                                                    	at org.hyperledger.ariesvcx.VcxUniFfiException$ErrorHandler.lift(vcx.kt:906)
                                                                                                    	at org.hyperledger.ariesvcx.VcxUniFfiException$ErrorHandler.lift(vcx.kt:905)
                                                                                                    	at org.hyperledger.ariesvcx.VcxKt.createInvitee(vcx.kt:1087)
                                                                                                    	at org.hyperledger.ariesvcx.MainActivity.setProfileHolder(MainActivity.kt:29)
                                                                                                    	at org.hyperledger.ariesvcx.MainActivity.access$setProfileHolder(MainActivity.kt:23)
                                                                                                    	at org.hyperledger.ariesvcx.MainActivity$onCreate$1$1$1$1.invoke(MainActivity.kt:41)
                                                                                                    	at org.hyperledger.ariesvcx.MainActivity$onCreate$1$1$1$1.invoke(MainActivity.kt:41)
                                                                                                    	at org.hyperledger.ariesvcx.HomeScreenKt$HomeScreen$1$1$1.invokeSuspend(HomeScreen.kt:47)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
                                                                                                    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [androidx.compose.ui.platform.MotionDurationScaleImpl@7db2491, androidx.compose.runtime.BroadcastFrameClock@698eff6, StandaloneCoroutine{Cancelling}@67f4ff7, Dispatchers.IO]
2023-08-29 13:18:04.311 14352-14418 Process                 org.hyperledger.ariesvcx             I  Sending signal. PID: 14352 SIG: 9
```

File: uniffi_aries_vcx/core/src/core/profile.rs [hunk 2]
```diff
@@ -16,9 +16,17 @@ impl ProfileHolder {}
 pub fn new_indy_profile(wallet_config: WalletConfig) -> VcxUniFFIResult<Arc<ProfileHolder>> {
     block_on(async {
         let wh = create_and_open_wallet(&wallet_config).await?;
-        let ph = 0;
-        // let profile = VdrtoolsProfile::init(wh, ph);
-        let profile = MockProfile {};
+        // let ph = 0;
```
Comment (gmulhearn) [inline uniffi_aries_vcx/core/src/core/profile.rs:19]: nit; can remove redundant comment

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        ))
+    }
+
+    Column(
+        modifier = Modifier.fillMaxSize(),
+        horizontalAlignment = Alignment.CenterHorizontally,
+        verticalArrangement = Arrangement.Center
+    ) {
+        Button(onClick = {
+            // How to persist this in the app state?
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt:49]: dont think this comment is relevant anymore

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            scope.launch(Dispatchers.IO) {
+                val profile = newIndyProfile(walletConfigState)
+                setProfileHolder(profile)
+                Handler(Looper.getMainLooper()).post {
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt:53]: why is this necessary? if you're trying to show the toast with the main context, i think using `withContext(Dispatchers.Main) { .. }` would be the more idiomatic way of doing this. (altho please correct me if this does not work)

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            text = { Text(decoded) },
+            confirmButton = {
+                TextButton(onClick = {
+                    scope.launch(Dispatchers.IO) {
+                        if (profileHolder != null) {
+                            connection?.acceptInvitation(
+                                profile = profileHolder,
+                                invitation = decoded
+                            )
+                            connection?.sendRequest(profileHolder, map["serviceEndpoint"] as String, map["routingKeys"] as List<String>)
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:70]: this is not the correct usage of `sendRequest`. `send_request` in aries_vcx takes in the service endpoint and routing keys that you own. Here you are passing in the endpoint and keys that the Inviter owns (bcus you're extracting them from the Inviter's invitation).

As such, it's probably best to just leave the serviceEndpoint and routingKeys as dummy inputs for now: e.g. `https://google.com` and `emptyList()`.

In your next PR we will worry about replacing those dummy inputs with real values (where the real values will involve the simple relay service i've mentioned before)

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+        AlertDialog(
+            onDismissRequest = { scannedQRCodeText = null },
+            title = { Text("Accept this invitation?") },
+            text = { Text(decoded) },
+            confirmButton = {
+                TextButton(onClick = {
+                    scope.launch(Dispatchers.IO) {
+                        if (profileHolder != null) {
+                            connection?.acceptInvitation(
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:66]: this is ok. But i think it would be preferable if this ScanScreen is logically unreachable with null `profileHolder` or `connection`.

You could do this by making the the "scan qr code" button on the home screen un-clickable until the profileHolder (and the connection) is not null

**Review by gmulhearn — CHANGES_REQUESTED** [review]
Nice, it's looking good. just some smallish comments, let me know when they're resolved or if any questions

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            walletName = "test_create_wallet_add_uuid_here",
+            walletKey = "8dvfYSt5d1taSd6yJdpjq4emkwsPDDLYxkNFysFD2cZY",
+            walletKeyDerivation = "RAW",
+            walletType = null,
+            storageConfig = null,
+            storageCredentials = null,
+            rekey = null,
+            rekeyDerivationMethod = null
+        ))
+    }
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/HomeScreen.kt:42]: this probably does not need to be a mutableState. i think just `val walletConfig = WalletConfig` is fine (since this variable will not change)

File: uniffi_aries_vcx/demo/app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -1,6 +1,12 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
+    <uses-feature
+        android:name="android.hardware.camera"
+        android:required="false" />
+    <uses-permission android:name="android.permission.CAMERA" />
+    <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/AndroidManifest.xml:9]: question: what does this do? / what is it needed for?

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/MainActivity.kt [hunk 9]
```diff
@@ -4,78 +4,78 @@ import android.os.Bundle
...
+                setProfileHolder = setProfileHolder,
+                profileHolder = profileHolder,
+                connection = connection
+            )
+        }
+
+        composable(Destination.QRScan.route) {
+            ScanScreen(
+                connection = connection,
+                profileHolder = profileHolder
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/MainActivity.kt:77]: hmmm i wonder if you could set these as non-nullable, since we know they logically cannot be null... e.g. `connect = connection!!, profileHolder = profileHolder!!`. and then ScanScreen could take non-nullable params

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    val context = LocalContext.current
+    val lifecycleOwner = LocalLifecycleOwner.current
+    val cameraProviderFuture = remember {
+        ProcessCameraProvider.getInstance(context)
+    }
+
+    scannedQRCodeText?.let { text ->
+        val encoded = Uri.parse(text)?.getQueryParameter("c_i")
+        val decoded =  String(Base64.decode(encoded, Base64.DEFAULT))
+        val map = jsonStringToMap(decoded)
```
Comment (gmulhearn) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/ScanScreen.kt:56]: no longer used, could remove function too

File: uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/vcx.kt
Comment (Patrik-Stas) [inline uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/vcx.kt:1]: This looks autogenerated, should this be provisioned in git?

Comment (gmulhearn) [reply uniffi_aries_vcx/demo/app/src/main/java/org/hyperledger/ariesvcx/vcx.kt:1]: As discussed in slack, this will be handled in next PR; https://github.com/hyperledger/aries-vcx/pull/968

**Review by gmulhearn — APPROVED** [review]
